### PR TITLE
A new journal web GUI

### PR DESCRIPTION
**Pull request by Samuel-BF**
# TL;DR

Here's a proposal for a new journal web GUI (journal-remote-gatewayd / browse) with fancy new features, but I'm a noob in C so it needs review.

Summary :
- Motivation / Goal
- New features in front-end
- New features in back-end (API)
- Things that should be especially reviewed / why is it coded that way
- Improvements that could be made

# Motivation / Goal

Reading logs, searching and investigating in logs is a daily task for most sysadmins. When you have a lot of systems to watch over, you'll tipically go with an [ELK stack](https://www.elastic.co/fr/what-is/elk-stack) to store, search, correlate and all the things. But if you have only 1 to ten servers, all looking the same and using systemd, deploy a full ELK stack seems a bit too much, just to process logs. It leads to duplicate storage (systemd-journal and logstash), adds work to convert logs from one mechanism to another and adds infrastructure needs.

With systemd, you already have a way to centralize logs (push mode with systemd-journal-remote and pull mode with systemd-journal-gatewayd) and have a basic browsing utility (systemd-journal-gatewayd / [browse.html](https://github.com/systemd/systemd/blob/master/src/journal-remote/browse.html)). This PR improves this browsing utility with several features, making it suitable for daily supervision and investigation.

# New features in front-end

- A follow mechanism : get the new entries in real time
- Filter on arbitrary journal fields
- Show side-by-side several entries
- Shareable views: just copy-paste the URL and send it to a teammate so he/she can have the same view as you (same entries in the list, same entries in the table below). These URL are predictable : an alerting mechanism could add a link to "this event" in the alert.
- Search by time of the event
- Show host name in the list of entries (hidden when there is only one host in the journal)
- From details pane (below the entries list), you can filter the entries list by any value you see.

I removed the original header with machine info.

journal-reader
# New features in back-end

[man page update](https://github.com/systemd/systemd/compare/master...Samuel-BF:journal-reader#diff-db865b2e6a8608e3ed8cde23e92f76b4)

- new `/fields` call : list field names in the journal (just like `journalctl --fields`)
- in `Range` header, you can filter by time range with `time=[from]:[to][[:num_skip]:[num_entries]` syntax, where `from` and `to` are microtimestamps.
- new `lines=n` GET parameter to get the last *n* entries of the journal, just like `journalctl --lines=n`. This is redundant with `Range` header, but it is necessary for the `follow` feature, since the [Server-sent events API](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) does not allow to specify custom headers in a request.
- "/" to "/browse" redirection is now relative (i.e. to "./browse"), allowing systemd-journal-gatewayd to work behind a reverse proxy (which is necessary if you want to protect this tool with authentication)

# Things that should be especially reviewed / why is it coded that way

I have a little experience in javascript, but that's the first time I write C code. So [my changes to journal-gatewayd.c](https://github.com/systemd/systemd/compare/master...Samuel-BF:journal-reader#diff-2d7e61049e387b011a9df20244a7922d) needs review in particular.

- **[Listing field names](https://github.com/systemd/systemd/compare/master...Samuel-BF:journal-reader#diff-2d7e61049e387b011a9df20244a7922dR751)** : I intially coded it with `MHD_create_response_from_callback` like in [request_handler_entries](https://github.com/systemd/systemd/blob/d99d1aaa21e216b2c2f520b58f847248407ba768/src/journal-remote/journal-gatewayd.c#L458) and [_fields](https://github.com/systemd/systemd/blob/d99d1aaa21e216b2c2f520b58f847248407ba768/src/journal-remote/journal-gatewayd.c#L604), but it was copy-paste code (which is dirty) and I thought that it was always quite a short list, so I buffered all at once in a temporary file and output it with `MHD_create_response_from_fd`. Is it OK ? If my understanding is good, `MHD_create_response_from_fd` will close fd and that will remove the file, without any need to do further cleaning : right ? Do you see any memory leak in this function ? (or somewhere else... or any other coding error)

- I [swapped request_parse_arguments and request_parse_range](https://github.com/systemd/systemd/compare/master...Samuel-BF:journal-reader#diff-2d7e61049e387b011a9df20244a7922dR586) so arguments are parsed first, which allows to give priority to `Range:` header over `lines` GET argument when both are present.

- For JS code, I tried to avoid XSS by using exclusively `new HTMLElement`, `htmlElement.setAttribute` and `document.createTextNode()` for displaying content, so that strings and tag names / attributes are never mixed (like they could be with inner/outerHTML attributes).

# Improvements that could be made

I won't implement these things for now, so you can merge this without waiting shiny things that I list here.

- **Boot filter** : implement a "/list-boots" URL which gives you something like `journalctl --list-boots`, so the Boot filter has a sensible presentation (for now, it displays only the  BOOT_ID, which is not very significant). But we need first to move `get_boots` function and `BootId` type from journalctl.c to sd-journal.c, and fix the the calculation of the boot list for a journal directory with logs from several hosts.
- ~~**Multi-value filters** : allow one to select several values for a filter (e.g. hostname), which makes an OR-ed filter (entries for host A or host B or ...). It's easy to code, but I'm very bad at designing interfaces and I don't know how to make this look "fancy".~~ done. 
- We could also try to implement arbitrary logical expressions, like in `journalctl`, but it seems far more complicated to design a user interface for that.
- ~~**favicon** : maybe something inspired by systemd logo ? the systemd logo with vertical symetry and replacing the dot with 2 vertical bars ("pause" symbol) ?~~ done :

 favicon

### CODE REVIEW

**Comment by Samuel-BF** [comment]
I present myself and as a JS coder and C noob, and lgtml.com [kicks me for my JS code](https://github.com/systemd/systemd/pull/17314/checks?check_run_id=1239194759). Shame on me.

Now fixed.

I added a screenshot and the suggestion of a favicon in initial presentation.

**Comment by Samuel-BF** [comment]
Hi, @poettering and @keszybz (main authors of browser/gateway) : would you accept to throw an eye on it ?

I would be pleased to work on #14625 and #17326, but I'd like a review on my C style first.

**Comment by Samuel-BF** [comment]
First version had :

- "follow" mode
- filter on arbitrary field
- seek at specific time
- select several entries at the same time

l modified the JS code to add :

- filter on several values at the same time (e.g. `SYSLOG_IDENTIFIER = sshd OR nginx`)
- support for sort of negative filters (like asked in #2720). "Sort of" because when selecting entries where `FIELD_A!=val1` and `FIELD_A!=val2`, you'll get only entries where `FIELD_A` has a value. And if an entry has several keys `FIELD_A`, one with `val1` and another with `val3` (for example), it will show up in the results.
- compliance with [Journal JSON Format](https://www.freedesktop.org/wiki/Software/systemd/json/)
- "download as CSV" button (it exports only entries currently shown in the Journal)
- binary to ASCII interpretation : in "Journal" pane, showing valid ascii characters and in "Details" view showing both hexadecimal and ascii version. I found 2 programs that emitted binary output on the console : Microsoft dotNet (which sends console control characters, probably for coloring output) and Microsoft SQL Server (which translates log messages but with Microsoft encodings).

Note that you don't have the same flexibility for filtering as with journalctl (no support for OR-ing complex expressions).

illustration of these features

**Comment by boucman** [comment]
Nice... I hope this gets merged.

File: src/journal-remote/journal-gatewayd.c [hunk 1]
```diff
@@ -490,10 +601,15 @@ static int request_handler_entries(
...
                 r = sd_journal_seek_cursor(m->journal, m->cursor);
+        else if (m->since_set && m->n_skip >= 0)
+                r = sd_journal_seek_realtime_usec(m->journal, m->since);
         else if (m->n_skip >= 0)
                 r = sd_journal_seek_head(m->journal);
+        else if (m->until_set && m->n_skip < 0)
+                r = sd_journal_seek_realtime_usec(m->journal, m->until);
         else if (m->n_skip < 0)
                 r = sd_journal_seek_tail(m->journal);
+
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: Please drop this unnecessary blank line. The error handling should be located just after the call(s).
Also, I think it is safer to add
```
else
        r = 0;
```

File: src/journal-remote/journal-gatewayd.c [hunk 2]
```diff
@@ -637,6 +753,63 @@ static int request_handler_fields(
...
+        assert(m);
+
+        r = open_journal(m);
+        if (r < 0)
+                return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "Failed to open journal: %m");
+
+        if (request_parse_accept(m, connection) < 0)
+                return mhd_respond(connection, MHD_HTTP_BAD_REQUEST, "Failed to parse Accept header.");
+
+
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: doubled empty lines.

File: src/journal-remote/journal-gatewayd.c [hunk 3]
```diff
@@ -637,6 +753,63 @@ static int request_handler_fields(
...
+}
+
+static int request_handler_field_names(
+                struct MHD_Connection *connection,
+                void *connection_cls) {
+
+        _cleanup_(MHD_destroy_responsep) struct MHD_Response *response = NULL;
+        RequestMeta *m = connection_cls;
+        int r, fd;
+        FILE *tmp;
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]:
```
_cleanup_close_ int fd = -1;
_cleanup_fclose_ FILE *tmp = NULL;
int r;
```

File: src/journal-remote/journal-gatewayd.c [hunk 4]
```diff
@@ -307,42 +345,99 @@ static int request_parse_range(
...
+        assert(connection);
+
+        range = MHD_lookup_connection_value(connection, MHD_HEADER_KIND, "Range");
+        if (!range)
+                return 0;
+
+        m->n_skip = 0;
+        if (startswith(range, "entries="))
+                return request_parse_range_entries(m, range);
+        else if (startswith(range, "time="))
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: Not a matter, but `else` is redundant.

File: src/journal-remote/journal-gatewayd.c [hunk 5]
```diff
@@ -307,42 +345,99 @@ static int request_parse_range(
...
+        if (!colon)
+                return -EINVAL;
+
+        if (colon - range > 0) {
+                _cleanup_free_ char *t;
+                t = strndup(range, colon - range);
+                if (!t)
+                        return -ENOMEM;
+
+                r = safe_atou64(t, &m->since);
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: Maybe, `parse_usec()` is better?

File: src/journal-remote/journal-gatewayd.c [hunk 6]
```diff
@@ -307,42 +345,99 @@ static int request_parse_range(
...
+        range += 5;
+        range += strspn(range, WHITESPACE);
+
+        colon = strchr(range, ':');
+        if (!colon)
+                return -EINVAL;
+
+        if (colon - range > 0) {
+                _cleanup_free_ char *t;
+                t = strndup(range, colon - range);
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c:389]: Please explicitly initialize `t` with `NULL`, and add a blank line after the variable declaration.
```
_cleanup_free_ char *t = NULL;

t = strndup(...);
```

File: src/journal-remote/journal-gatewayd.c [hunk 7]
```diff
@@ -307,42 +345,99 @@ static int request_parse_range(
...
+        if (!colon)
+                until = strdup(range + 1);
+        else {
+                r = request_parse_range_skip_and_n_entries(m, colon);
+                if (r < 0)
+                        return r;
+
+                until = strndup(range + 1, colon - range - 1);
+        }
+
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: Please drop this empty line. The error handling must be located right after the function call.

File: src/journal-remote/journal-gatewayd.c [hunk 8]
```diff
@@ -307,42 +345,99 @@ static int request_parse_range(
...
+        if (!until)
                 return -ENOMEM;
 
-        m->cursor[strcspn(m->cursor, WHITESPACE)] = 0;
-        if (isempty(m->cursor))
-                m->cursor = mfree(m->cursor);
+        until[strcspn(until, WHITESPACE)] = 0;
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: Is this necessary?

File: src/journal-remote/journal-gatewayd.c [hunk 9]
```diff
@@ -307,42 +345,99 @@ static int request_parse_range(
...
+        if (!isempty(until)) {
+                r = safe_atou64(until, &m->until);
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: `parse_usec()`?

File: src/journal-remote/journal-gatewayd.c [hunk 10]
```diff
@@ -307,42 +345,99 @@ static int request_parse_range(
...
+                if (r < 0) {
+                        free(until);
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: Please drop this and use `_cleanup_free_` attribute.

File: src/journal-remote/journal-gatewayd.c [hunk 11]
```diff
@@ -307,42 +345,99 @@ static int request_parse_range(
...
+                        return r;
+                }
+
+                m->until_set = true;
+        }
+        free(until);
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: Please drop this.

File: src/journal-remote/journal-gatewayd.c [hunk 12]
```diff
@@ -307,42 +345,99 @@ static int request_parse_range(
         if (!colon)
                 m->cursor = strdup(range);
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c:353]: If `request_parse_range_entries()` is called twice, `m->cursor` is leaked. Maybe it is better to add `m->cursor = mfree(m->cursor);`

File: man/systemd-journal-gatewayd.service.xml [hunk 13]
```diff
@@ -214,9 +221,13 @@
     <para>
       <option>Range: entries=<replaceable>cursor</replaceable>[[:<replaceable>num_skip</replaceable>]:<replaceable>num_entries</replaceable>]</option>
     </para>
+    <para>
+      <option>Range: time=[<replaceable>since</replaceable>]:[<replaceable>until</replaceable>][[:<replaceable>num_skip</replaceable>]:<replaceable>num_entries</replaceable>]</option>
+    </para>
 
     <para>where
       <replaceable>cursor</replaceable> is a cursor string,
+      <replaceable>since</replaceable> and <replaceable>until</replaceable> are unsigned integers (microtimestamps),
```
Comment (yuwata) [inline man/systemd-journal-gatewayd.service.xml]: `unsigned integers` -> `timestamp`, and use `parse_usec()` is better.

**Review by yuwata — CHANGES_REQUESTED** [review]
Not fully reviewed yet. Only C part.

File: src/journal-remote/journal-gatewayd.c (hunk 8 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: With this line : `range: time=1618136100000000:1618136400000000   :5` is accepted, while it is not without (because of the spaces between the second micro-ts and the colon). There's no interest in accepting values outside of the spec, so I removed it.

File: src/journal-remote/journal-gatewayd.c (hunk 3 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: It breaks the code: in the logs I get "microhttpd: Failed to send data in request for '/fields'.".

Note that there's a design decision behind this code : 
> Listing field names : I intially coded it with MHD_create_response_from_callback like in request_handler_entries and _fields, but it was copy-paste code (which is dirty) and **I thought that it was always quite a short list**, so I buffered all at once in a temporary file and output it with MHD_create_response_from_fd. Is it OK ? 

It's not always a short list, and may take time to execute. So maybe the first solution (copy-paste) is better, or maybe another level of abstraction : a function that takes one or several function as argument(s) and would be suitable for running `request_handler_entries`, `_fields` and `_field_names` (replacing `request_reader_entries` and `request_reader_fields`). Do you have any opinion on that point ?

File: src/journal-remote/journal-gatewayd.c (hunk 4 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: Micro-optimization : when first condition is met, the second is not evaluated (but maybe the compiler handles it automatically ?). If it's more clear without else, I remove it.

File: src/journal-remote/journal-gatewayd.c (hunk 5 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: I've not found parse_usec function, only a target for jumps in [basic/time-util.c](https://github.com/systemd/systemd/blob/main/src/basic/time-util.c#L822). I'm sorry, I don't know how to use it (I'm really beginning with C by this code). Could you precise your hint ?

File: src/journal-remote/journal-gatewayd.c (hunk 6 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c:389]: done.

File: src/journal-remote/journal-gatewayd.c (hunk 7 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: done.

File: src/journal-remote/journal-gatewayd.c (hunk 10 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: done.

File: src/journal-remote/journal-gatewayd.c (hunk 11 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: done.

File: src/journal-remote/journal-gatewayd.c (hunk 12 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c:353]: Ok ! done.

File: src/journal-remote/journal-gatewayd.c (hunk 2 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: done

File: src/journal-remote/journal-gatewayd.c [hunk 14]
```diff
@@ -283,70 +296,149 @@ static int request_parse_accept(
...
-                        if (m->n_entries <= 0)
-                                return -EINVAL;
+        range += 8;
+        range += strspn(range, WHITESPACE);
 
-                        m->n_entries_set = true;
-                }
+        colon = strchr(range, ':');
+        if (!colon) {
+                m->cursor = mfree(m->cursor);
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c:352]: Please move this `mfree()` before `colon = strchr(range, ':');` Otherwise, `m->cursor` may leak in the `else` branch.

File: src/journal-remote/journal-gatewayd.c [hunk 15]
```diff
@@ -283,70 +296,149 @@ static int request_parse_accept(
...
                 return -ENOMEM;
 
-        m->cursor[strcspn(m->cursor, WHITESPACE)] = 0;
         if (isempty(m->cursor))
                 m->cursor = mfree(m->cursor);
 
         return 0;
 }
 
+
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: doubled blank lines.

File: src/journal-remote/journal-gatewayd.c [hunk 16]
```diff
@@ -396,6 +488,21 @@ static mhd_result request_parse_arguments_iterator(
                 return MHD_YES;
         }
 
+        if (streq(key, "lines")) {
+                if (isempty(value)) {
+                        return MHD_NO;
+                }
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: Coding style: please drop unnecessary `{}`

File: src/journal-remote/journal-gatewayd.c [hunk 17]
```diff
@@ -396,6 +488,21 @@ static mhd_result request_parse_arguments_iterator(
...
+
+                r = safe_atoi64(value, &m->n_skip);
+                if (r < 0) {
+                        m->argument_parse_error = r;
+                        return MHD_NO;
+                }
+
+                m->n_skip *= -1;
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c:506]: Hmm?? Why this is necessary?

File: src/journal-remote/journal-gatewayd.c [hunk 18]
```diff
@@ -637,6 +751,63 @@ static int request_handler_fields(
...
+}
+
+static int request_handler_field_names(
+                struct MHD_Connection *connection,
+                void *connection_cls) {
+
+        _cleanup_(MHD_destroy_responsep) struct MHD_Response *response = NULL;
+        RequestMeta *m = connection_cls;
+        int fd = -1;
+        FILE *tmp = NULL;
```
Comment (yuwata) [inline src/journal-remote/journal-gatewayd.c]: Hmm, actually my previous suggestion is not correct, but these should be closed on failure...

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/journal-remote/journal-gatewayd.c (hunk 5 above)
Comment (yuwata) [reply src/journal-remote/journal-gatewayd.c]: Ah, sorry, `parse_sec()`.

File: src/journal-remote/journal-gatewayd.c (hunk 4 above)
Comment (yuwata) [reply src/journal-remote/journal-gatewayd.c]: We usually drop redundant `else`. I hope compiler handles both cases well.

Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: Done. And I was wrong : in `if A { return X; } else if B { ... } ` the B is not evaluated when A is true !

File: src/journal-remote/journal-gatewayd.c (hunk 14 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c:352]: Done

File: src/journal-remote/journal-gatewayd.c (hunk 15 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: Fixed

File: src/journal-remote/journal-gatewayd.c (hunk 16 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: Fixed

File: src/journal-remote/journal-gatewayd.c (hunk 17 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c:506]: It determines [if we seek head or tail](https://github.com/systemd/systemd/pull/17314/commits/cdc2b0e58bac0baf042c1c7d8a8d9ed6f1e7713d#diff-732cab1fcb807574656bfc1b912fe5e3b41ecba76bdbda6bfb550bf3587f4681R605). When n_skip > 0, we seek on journal head, when < 0 we seek on tail. And the lines GET argument is only used for recent logs ("Show the most recent journal events and limit the number of events shown" as I put it in the doc).

File: src/journal-remote/journal-gatewayd.c (hunk 18 above)
Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: Right ! But I'm not sure it's the right way to code it. Other options :

- On branch ["copypaste"](https://github.com/Samuel-BF/systemd/blob/journal-reader-copypaste/src/journal-remote/journal-gatewayd.c#L761) I implemented `field_names` requests like for `fields_request` : [request_handler_field_names](https://github.com/Samuel-BF/systemd/blob/journal-reader-copypaste/src/journal-remote/journal-gatewayd.c#L834) calls [MHD_create_response_from_callback](https://github.com/Samuel-BF/systemd/blob/31a52c03253f346af5156d6c179cfe9ff985f2df/src/journal-remote/journal-gatewayd.c#L854) with a (quite convoluted) function that reads the next field name and puts into MHD buffer.

This way, you have 3 functions ([request_handler_entries](https://github.com/Samuel-BF/systemd/blob/journal-reader-copypaste/src/journal-remote/journal-gatewayd.c#L566), [request_handler_fields](https://github.com/Samuel-BF/systemd/blob/journal-reader-copypaste/src/journal-remote/journal-gatewayd.c#L721) and [request_handler_field_names](https://github.com/Samuel-BF/systemd/blob/31a52c03253f346af5156d6c179cfe9ff985f2df/src/journal-remote/journal-gatewayd.c#L834)) which are very similar. Is it better ? (same logic everywhere)

- We could merge the 3 functions by adding members to RequestMeta objects. That would be shorter, would avoid copypaste, but would also add another level of complexity. Do you think it remains manageable if we do this way ? (I'm OK to (try to) code it this way if you think it's worth trying !)

And, btw : thanks a lot for your review and the time/energy taken for this.

Comment (Samuel-BF) [reply src/journal-remote/journal-gatewayd.c]: Fixed by adding `TAKE_PTR(tmp)` after `TAKE_FD(fd)`.

If I'm correct, in the everything-works case : 

-  MHD_create_response_from_fd will [close fd upon completion](https://www.gnu.org/software/libmicrohttpd/manual/html_node/microhttpd_002dresponse-create.html#index-MHD_005fcreate_005fresponse_005ffrom_005ffd)
- this will free (close ?) the `tmp` file stream as a consequence (man 2 close : "the resources associated with the open file description are freed")

And if there's an error somewhere:

- compiler inserted `close` and `fclose` calls everywhere (thanks to `_cleanup_f?close_`). It is redundant (we could remove either `_cleanup_close_`, either `_cleanup_fclose_`), as closing the fd will free the FILE * stream and the other way round, but ... it's more clean anyway.

Am I right ?

I also coded [the refactoring](https://github.com/Samuel-BF/systemd/blob/journal-reader-refactor/src/journal-remote/journal-gatewayd.c), using init/iterator functions in RequestMeta for `/fields`, `/fields/xxx` and `/entries` endpoints. The code is shorter than in the copypaste version, but harder to follow and understand. So the 3 options are : 
- this one (here), where we use a different logic for field and field names (a bit inconsistent)
- [copy-paste](https://github.com/Samuel-BF/systemd/blob/journal-reader-copypaste/src/journal-remote/journal-gatewayd.c#L761) which uses the very same logic for both, but is a bit redundant.
- [refactor](https://github.com/Samuel-BF/systemd/blob/journal-reader-refactor/src/journal-remote/journal-gatewayd.c) which is shorter, but less readable (and imho a bit more messy)

**Comment by Samuel-BF** [comment]
> Not fully reviewed yet. Only C part.

Before you start review JS, I have a question and 2 remarks : 

- Would it be OK to [split browse.html in 3 files (html/js/css)](https://github.com/Samuel-BF/systemd/tree/journal-reader-browsesplit) ? It makes the code longer in lines (+140) but shorter in characters (-20k, indentation),  and allows a tighter [Content-Security-Policy HTTP header](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) sandbox.
- The code is not very sound about error handling. For my own use, it was OK, but since you consider sharing it with a larger audience, I'll try to improve this aspect 
- Small presentation of the code : 
   - we store all information in the DOM
   - we serialize the control form into URL (`_setBrowserURL`), and deserialize from URL on startup (`_initFiltersFromURL`), allowing to share "view" of the journal
   - Journal fields are treated as multi-value Filters, except : UID/GID/PID (-> number filters), Message (-> text filter), Cursor (-> simple field, no filter), Realtime Timestamp (-> defines the "since" argument), Priority (-> slider), Timestamps (-> simple fields)
   - Request to `/entries` are built from filters in `reloadJournal` (calling `valueToEntriesQuery` member function for each filter) and from `_getRange` for "static" queries or only from the `_form.lines` value for queries in follow mode.
   - ... I'm here if you have any question or you notice something unclear.

**Comment by Samuel-BF** [comment]
> The code is not very sound about error handling. For my own use, it was OK, but since you consider sharing it with a larger audience, I'll try to improve this aspect

It's a bit better now : 
- we only show errors to user when they have impact on entries shown. e.g if we fail to load `/fields/SYSLOG_IDENTIFIER` (field values for SYSLOG_IDENTIFIER), it's OK and we will retry if the user asks for it by clicking the select arrow, unless the user selected "≠" for SYSLOG_IDENTIFIER (thus we need all existing values to compute the query for `/entries`), in which case we display "Server (or connection) error. retry ?"
- when we see "new" fields in data (before getting the results of `/fields` request, or after getting this result but for a field that was not present in `/fields` response - .e.g when a event introduced a new field), we add this field to the list. Likewise, when we see new value for a field, we add it to the list.

Actually, error handling makes `browse.html` code work quite well with current version of `systemd-journal-gatewayd` (before this PR) : field list is inferred from events and not from `/fields`, follow mode doesn't work and you can't seek by date-time, but outside of these limitations, it works.

**Comment by Samuel-BF** [comment]
Last commit : 
- fix bugs in display (appError box was displayed on startup,  loading arrow didn't stop in follow mode, and selected entries were not cleared when toggling follow mode)
- add a Content-Security-Policy header (a bit less strict than in [browsesplit branch](https://github.com/Samuel-BF/systemd/tree/journal-reader-browsesplit)

I published [a demo](https://lofolomo.ovh/journal/browse/) online (user: demo / password: lofolomo), as part of a [prototype for a log-to-alert pipeline](https://gitlab.com/samuelbf/lofolomo).

**Comment by Samuel-BF** [comment]
This pull request is sitting here for quite a long time and still works (i rebased it on master). The C part has been reviewed (thanks @yuwata !) and the JS part not yet.

The "GUI part" (JS code) can be distributed independently from systemd.

The C part allows to seek by time (fixing #4883), list field values and adds a Content-Security-Policy header.

My proposal is to strip off JS part of this pull request to get C part merged. Would you approve ?

**Comment by yuwata** [comment]
Oh, it seems something you made mistake while rebasing, as you can see thousands of commits in this PR now. Please re-rebase this PR.

**Comment by Samuel-BF** [comment]
> Oh, it seems something you made mistake while rebasing, as you can see thousands of commits in this PR now. Please re-rebase this PR.

Oops, I got hit by the master/main rename. Fixed.

I've also repaired [the GUI demo](https://lofolomo.ovh/journal/browse) (credentials: demo / lofolomo).

**Comment by medhefgo** [comment]
We already have `docs/favicon.{png,svg}`, imho there's no need to for another one.

**Comment by medhefgo** [comment]
Here are some thoughts regarding usability. Not sure how difficulty any of these are to implement and whether you're interested in working on them (my web-fu is basically frozen in the web 2.0 days):
 - Half the page is used for the filters / log output control UI, which for a log viewer is simply bad. The primary objective is to look at logs and should therefore take over most of the view. It would be nicer if the controls would be compacted/unified and also have it prefer horizontal space.
 - The output control buttons are terribly misaligned for me. Looking at web dev tools, it seems to be caused by different fonts being selected: A mix of `Hack` and `Noto Symbols`: image
 - Haven't really tested this, but support for css dark mode would probably be nice.
 - Looking at some of the css, I see some font-sizes fixed at 7px. CSS provides pre-defined font-sizes which should be preferred (despite browser being eternally stuck at 90s when it comes to default fonts/font sizes). Also, some of the margins/paddings seem overly large, which gets especially silly when you zoom in.
 - And generally, I dislike how it basically re-implements the browser's scrollbar with the UI controls instead of using infinite scrolling (or "load older/newer logs"/"follow mode" buttons at the beginning/end of the list). The UI for filters etc. would be a floating bar at the top that's always visible.

**Comment by Samuel-BF** [comment]
Thanks a lot for your external sight and advice !
> my web-fu is basically frozen in the web 2.0 days

Me too. And I'm not a designer at all. I tried to reuse existing layout (modified a bit), but I take your advice as a "go" for (trying to) make something more modern.

> The primary objective is to look at logs

For me, the primary goal is to crawl/filter/analyze events using journal fields (one of the most underused feature of systemd, imho), that's why it take so much place. But there's probably a better way to present it without taking half the page !

> The output control buttons are terribly misaligned for me. 

Ugly, indeed. I'll try to fix that !

> Looking at some of the css, I see some font-sizes fixed at 7px.

OK ! I got converted by webdev to "em" units, then i learned that it was always basically a multiple of pixels and change nothing at the behavior then ... (webdev : lots of strict rules about "the right way" that changes every morning !)

> And generally, I dislike how it basically re-implements the browser's scrollbar with the UI controls instead of using infinite scrolling (or "load older/newer logs"/"follow mode" buttons at the beginning/end of the list). The UI for filters etc. would be a floating bar at the top that's always visible.

That sounds great ! I'll try to make it that way.

> whether you're interested in working on them

Yes, but expect no ETA.

**Comment by Paiusco** [comment]
@Samuel-BF have you considered spliting this PR into 2? I've been looking (and had to develop myself an app to do similar things as the gatewayd with since/until would) for this feature for a while.

I think that splitting it and letting the since/until improvements being merged and solving an issue from 2016 would be very much appreaciated :)
